### PR TITLE
Explore variable declares, func pointers, parameter improvements, and more

### PR DESCRIPTION
**Pull request by matter123**
This PR implements a fix for #26 and #134, as well as a partial fix for #71.

Patterns added:
 - `:qualified_type` a single type with an optional template call and preceding scope, avoids non-type tokens
- `:declaration` a qualified type followed by a variable name and zero or more square brackets, avoids preceding `(`
- `:declarations` same as declaration but allows multiple variables;
- `:type_alias` a `using` type alias (e.g. `using result_type = size_t;`)
- `:function_pointer` a function pointer declaration, optional name, any number of parameters.
- `:function_parameters` a single function parameter

### CODE REVIEW

**Comment by matter123** [comment]
Screenshot from 2019-04-26 20-46-23

**Comment by matter123** [comment]
this does not support yet multiple declarations where one of the variables is immediately succeeded by `{`.
e.g.
```c++
int a, b{5}`
```
Screenshot from 2019-04-26 20-55-37

note how 5 has no syntax highlighting.

**Comment by matter123** [comment]
Screenshot from 2019-04-27 09-52-59

Fixed the issue with no comment highlighting and the lack of univeral initalization highlighting.

**Comment by matter123** [comment]
currently, the pattern causes the union scope not to end.

**Comment by matter123** [comment]
This is currently unsuitable for merging.

Regression:
-  functions with multiple parameters
- asm blocks
 - TODO

**Comment by matter123** [comment]
function definitions are particularly troublesome as empty functions do not contain semicolon to stop the declarations;

**Comment by matter123** [comment]
A side effect of this is using declarations inside structs have the alias marked as a type.
i.e `using type = T` marks type as `entity.name.type`

**Comment by matter123** [comment]
the following has bad highlighting
```c++
template <class T = value_type>
std::enable_if_t<std::is_trivially_move_constructible_v<T> &&
			std::is_trivially_move_assignable_v<T> &&
			std::is_contiguous_iterator<BidirIt>::value &&
			std::is_contiguous_iterator<OutputIt>::value>
operator()(Allocator &alloc, BidirIt first, BidirIt last,
	   OutputIt dfirst, std::priority_tag<6>) {
		size_t size = std::distance(first, last) * sizeof(value_type);
		std::memmove(addr(dfirst), addr(first), size);
		std::destruct<BidirIt, Allocator,
		              std::destruct_after::move_construct>(alloc, first,
		                                                   dfirst);
}
```
Screenshot from 2019-04-27 22-38-34
(ignore the bold)

**Comment by matter123** [comment]
I think the issue in the last comment can be solved by limiting declarations to be the first thing on a line. I'll try that in the morning.

**Comment by matter123** [comment]
:scope_resolution was moved to the last item in body_includes for struct/class/union.

**Comment by matter123** [comment]
should the `*` or `&` in parameters & declarations such as `const char *argv` be `entity.name.type.modifier` or `variable.other.type_modifier` (i.e. are they part of the type or variable)? Should be answer be different if there is no named parameter (e.g `int strcmp(const char *, const char *);`)?

**Comment by jeff-hykin** [comment]
I'm really excited to look into this PR, I'm going to do it last though since I think it has the largest changes.

**Comment by jeff-hykin** [comment]
Okay, I'm going to work on getting the tests passing for this and get it merged in 👍

**Comment by matter123** [comment]
You might want to hang on, I'm going to submit a PR to fix the sorting of the specs.

**Comment by jeff-hykin** [comment]
> should the `*` or `&` in parameters & declarations such as `const char *argv` be `entity.name.type.modifier` or `variable.other.type_modifier` (i.e. are they part of the type or variable)? Should be answer be different if there is no named parameter (e.g `int strcmp(const char *, const char *);`)?

This is tough, I actually think we should keep it as `keyword.operator` for now and maybe ask alexr about it. It definitely is a modifier, and it would be nice if it was a similar color to the type. I'll get back to you on this.

**Comment by jeff-hykin** [comment]
and okay I'll wait

**Comment by jeff-hykin** [comment]
Screen Shot 2019-05-08 at 10 38 54 PM
Putting this here to keep track of it.

**Comment by jeff-hykin** [comment]
Tomorrow I'll work on trying to get the function pointer syntax working inside function parameters. After that I think it'll just be a matter of updating all the tests.

**Comment by matter123** [comment]
Are you referring to the templated function syntax `template<void(void)`> or something else? `foo(void(*)(void))`?

**Comment by jeff-hykin** [comment]
Just the `function<void(void)>` but hopefully it should work with everything.

**Comment by matter123** [comment]
so a templated function syntax supports parameter names but not default arguments https://godbolt.org/z/24gB-m.

**Comment by matter123** [comment]
code

**Comment by matter123** [comment]
qualified_type might be overly broad in labeling the entire `function<void(void, usertype userval)>` as a type

**Comment by matter123** [comment]
With `function_pointer` again being marked as a parameter. Let's focus again on merging. For most purposes where your function type does not have a name, it is correct, and when it does have a name. The colorization is no worse than it is on the master branch.
Screenshot from 2019-05-09 10-21-30

**Comment by jeff-hykin** [comment]
Okay I agree, I'll try to get it merged in by tonight. Thanks for already getting it working with function  pointer support. In the most recent picture, are the parameter types being matched at? ( the `Event` or the `function` )

**Comment by matter123** [comment]
Sorry I wasn't clear the most recent image 3 hours ago is 1.8.13. The image 14 hours ago is the most recent commit.

**Comment by jeff-hykin** [comment]
I'm thinking after the merge, we can make a range to match the type+variable. The the range can just include the function pointer syntax or any complicated templating/scope resolution.

**Comment by jeff-hykin** [comment]
And ah okay, yeah I'm not too concerned with the named parameter inside the function pointer syntax.

**Comment by matter123** [comment]
The accidental premature merge of Explore-#71 and revert meant that `git merge master` no longer produced a correct merge. This has been rectified. `cpp/generate.rb` now has a successful merge of Explore-#71 and master

**Comment by matter123** [comment]
it looks like issue 11 has a regression `some_var` is marked as `entity.name.type`

**Comment by matter123** [comment]
With regards to `some_var` we can keep existing behavior where a comment is considered a type. Or we can try walking backwards past the comment and fall back to existing behavior if the comment is multi-line. Ruby supports variable length lookbehinds right?

**Comment by matter123** [comment]
Another regression (from an earlier version of this PR)
Screen Shot 2019-05-10 at 09 03 23
The first `compound<typename templated::type *>::type value;` is missing proper highlighting for `typename`, `type`? and `*`. This is not because of the overly broad labeling in qualified_type. The tokens are missing any of their correct scope.

**Comment by matter123** [comment]
I resolved `some_var` by keeping the existing `probably_a_parameter heuristic`. if an identifier follows a block comment it is a parameter.

**Comment by matter123** [comment]
Regression in issue 16.
Screen Shot 2019-05-10 at 11 10 16

**Comment by matter123** [comment]
`qualified type` needs to be aware of multi-word types such as `long double`, `long long int`, `long int` `short int

Added:
```c++
void multi_word_types(signed char, unsigned char, short int, long int, long long,
                      long double, struct a, enum b);
``` to test/fixtures/071.cpp

**Comment by matter123** [comment]
`.zeroOrMoreOf(/unsigned|signed|short|long/.then(@spaces))` turning into `(?:unsigned|signed|short|long\s+)* is a bug right?

**Comment by jeff-hykin** [comment]
Thanks for handling the merging of master. Sadly variable length lookbehinds aren't supported. I think the `compound<typename templated::type *>::type value;` case should probably be changed before merging it in.

**Comment by matter123** [comment]
The tests up to (but not including) issues/055.cpp have been fixed.

**Comment by jeff-hykin** [comment]
templating got messed up a little bit, I'm not sure when this happened though.
function templating messed up

**Comment by matter123** [comment]
I think its because `function_parameters` includes `<>`

**Comment by jeff-hykin** [comment]
Just made a commit for this:
same-line declares issue

**Comment by matter123** [comment]
Ok, that should probably be added to `issues/071.cpp`

**Comment by matter123** [comment]
How did < and > ever get marked as keyword.comparison? None of the template patterns do so.

**Comment by jeff-hykin** [comment]
I think it is from the $base context. In the old grammar all of the `<>`'s were always marked as comparison.

**Comment by matter123** [comment]
so adding `:operators` to `template_call_context` should do it?

**Comment by jeff-hykin** [comment]
Probably, I'm not sure exactly what changed from before though.

**Comment by matter123** [comment]
So I changed in scope resolution. `template_call.without_numbered_capture_groups` to plain `templare_call`

**Comment by jeff-hykin** [comment]
Screen Shot 2019-05-10 at 3 27 29 PM

**Comment by jeff-hykin** [comment]
We might be able to fix that last one by restricting where a function definition can occur. For example, the beinging/root of a file, inside a struct, inside a class etc.

**Comment by matter123** [comment]
So the difference in color is because `characters_in_template_call` was expanded to include `()` to match `<void(void)>`. Removing that makes all four angle brackets white for me.

**Comment by matter123** [comment]
> We might be able to fix that last one by restricting where a function definition can occur. For example, the beinging/root of a file, inside a struct, inside a class etc.

Yeah, It usually is. inside a function, for instance, things are highlighted normally. I'm guessing that lambda bodies are using the root scope.

**Comment by jeff-hykin** [comment]
So it's including `:function_context_c` which is including `:$initial_context`. I'll work on a fix for that

File: cpp/examples.cpp [hunk 1]
```diff
@@ -150,7 +150,6 @@
 //
     auto a = new int(5);
     delete a;
-    new (&a_storage_of_callable.callable) type(forward<Callable>(a_callable));
```
Comment (matter123) [inline cpp/examples.cpp]: What makes this invalid?

**Comment by jeff-hykin** [comment]
So that fixed the lambda's one.
Screen Shot 2019-05-10 at 4 08 58 PM

**Comment by matter123** [comment]
Not quite, I reverted the change that added `(` and `)` to characters allowed in a template call. That fixes that but breaks `<void(void)>`

**Comment by jeff-hykin** [comment]
For some reason though the first parameter in lambda captures are different
Screen Shot 2019-05-10 at 4 10 24 PM

**Comment by matter123** [comment]
I think its the `...`

**Comment by jeff-hykin** [comment]
I think qualified type might need to be a range :/

**Comment by matter123** [comment]
it probably should be, but can't

**Comment by jeff-hykin** [comment]
In general yeah, but I think for function parameters it might be able to, and it might be possible for declarations inside root/class/struct/namespace areas too.

**Comment by matter123** [comment]
Declarations is a range, as is function prameters.

File: cpp/examples.cpp (hunk 1 above)
Comment (jeff-hykin) [reply cpp/examples.cpp]: I don't think `new` can be used in that way in the root context.

Comment (matter123) [reply cpp/examples.cpp]: Ok, `example.cpp` needs quite a bit of work in that regard then. bare numbers aren't allowed in the root context as well.

**Comment by matter123** [comment]
Screen Shot 2019-05-10 at 14 23 53
Fixed lambda capture.

**Comment by jeff-hykin** [comment]
I think issues like the `compound<typename templated::type *>::type value;` and the `function<void(void)>` could be solved if they used a range-based qualified type instead of a pattern based one.

Templated function calls will probably always need a pattern based qualified type though.

**Comment by jeff-hykin** [comment]
If we can identify the area that a qualified type is supposed to be in then I think we can create a qualified range pattern that will correctly parse any type no matter how complex it is.

**Comment by jeff-hykin** [comment]
Screen Shot 2019-05-10 at 4 15 44 PM

**Comment by jeff-hykin** [comment]
It might be best to split this merge up into smaller changes, and then merge the smaller changes in one at a time.

**Comment by matter123** [comment]
Either skip past the block comment or end the parameter.

Yeah, that's probably a good idea. This is getting pretty complicated.

**Comment by matter123** [comment]
function_pointer is probably the easiest to cut out ill start there.

**Comment by jeff-hykin** [comment]
Sounds Good 👍 
We can probably do
- function pointer
- function definition range (old parameter matching)
- function call
- function definition (new parameter matching)
- variable declarations

**Comment by jeff-hykin** [comment]
I should be free all of today/tonight, I won't make any changes to master so that its easy to merge this in and I'll check periodically to merge pieces of this in. 

If you want to delegate anything let me know, I can probably at least update the tests with each merge.

**Comment by jeff-hykin** [comment]
I think the next thing to work on would probably be to seperate contexts a bit.

Theres a kind a "declaration" context that can't have if statements and other runtime code. I think thats root/class/struct/namespace and maybe some other places. Then there's function body context that can't have function definitions, but can have pretty much everything else. And then there's the parameter context that is just a list of types/parameters.

I think separating these out will help with the declarations. Let me know if you want me to implement them

**Comment by matter123** [comment]
Sure, go ahead and separate the contexts.

**Comment by jeff-hykin** [comment]
This is a lot more difficult than I expected it to be. I'm making good progress, but there are some challenges like the C dependencies. I'm also concerned about objective-c.

Since most everything imports from $initial_context right now, everything kind of needs to split between `declares` (root/class/struct/namespace) and `runtime` (functions, control flow), but it means nothing is inheriting from $inital_context.

All of the C++ changes have been made successfully. I tested, which was fantastic, there's absolutely no way this could've been done without the tests. @matter123  I could use your help with the 017 (universal characters) test, and I had to change 050.cpp since the switch statement needed to be inside of a main function to be valid. I'm not sure how to generate the `050.cpp.json` from the `050.cpp` file. The changes are on the `context_improvements` branch.

I'll conceptually figure out what to do for tonight in regard to the objective-c and C dependence. If there's no good solution, it might be best to just save the context changes for a later time and move forward with the other declaration changes.

**Comment by matter123** [comment]
Yeah, I can work on `issues/017`.

`npm run generate-specs -- issues/050.cpp` will regenerate the spec file for `issues/050.cpp`

**Comment by matter123** [comment]
You actually need a third context.

Scoped function definitions are not valid inside class definitions.
And unscoped constructors are not valid outside class definitions.

See https://godbolt.org/z/eWEns_

**Comment by matter123** [comment]
I'm going to fix 017 on a separate branch because it modifies `entity.name.function` on line 3046 and I want clean tests for that.

**Comment by jeff-hykin** [comment]
Yeah I'm thinking there will be a bunch down the line. An eval context to fix the `(int a)` being invalid, a conditional context for if statements, for `for` loops and while loops, etc. I think I'm mostly done for tonight though.

**Comment by matter123** [comment]
Does
> function definition range (old parameter matching)

Refer to #80?
If so thats blocked by #127 and #163.

**Comment by jeff-hykin** [comment]
Lets close this PR (reopen it if you feel differently) since many of the elements (function pointers, function definitions, qualified_type, type_alias) have been merged in. I think there's a few things we need to cover first like the objective C/C++ issue, and correcting the contexts that we should do before adding the `:declaration` and `:declarations` patterns.

Def. keep the code from this branch though to use it in future PR's.

**Pull request closed without merging**